File: Assets/Imports/EckyECS/Util/BitVector.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using System.Text;
using UnityEngine;
using UnityEngine.Assertions;

[StructLayout(LayoutKind.Sequential, Pack = 1)]
/** 
 * "string" of bits indicating wheether something is represented by this vector or not
 * 
 * Uses multiple ints as byte[] is for some reason still managed, even with MarshalAs. See
 * see https://limbioliong.wordpress.com/2011/06/29/passing-managed-structures-with-arrays-to-unmanaged-code-part-1/
 * for more info
 */
public unsafe struct BitVector
{
    private const int IntCount = 4;
    private const int IntSize = 32;

    // since mem layout is sequential we can use ptrs here
    // its still not pretty though
    private int Data0;
    private int Data1;
    private int Data2;
    private int Data3;

    public readonly BitVector And(BitVector Other)
    {
        BitVector NewVector = new();
        NewVector.Data0 = (Data0 & Other.Data0); 
        NewVector.Data1 = (Data1 & Other.Data1); 
        NewVector.Data2 = (Data2 & Other.Data2); 
        NewVector.Data3 = (Data3 & Other.Data3); 
        return NewVector;
    }

    public readonly BitVector Or(BitVector Other)
    {
        BitVector NewVector = new();
        NewVector.Data0 = (Data0 | Other.Data0);
        NewVector.Data1 = (Data1 | Other.Data1);
        NewVector.Data2 = (Data2 | Other.Data2);
        NewVector.Data3 = (Data3 | Other.Data3);
        return NewVector;
    }

    public readonly BitVector Subtract(BitVector Other)
    {

        BitVector NewVector;
        NewVector.Data0 = Mathf.Max(Data0 - Other.Data0, 0);
        NewVector.Data1 = Mathf.Max(Data1 - Other.Data1, 0);
        NewVector.Data2 = Mathf.Max(Data2 - Other.Data2, 0);
        NewVector.Data3 = Mathf.Max(Data3 - Other.Data3, 0);
        return NewVector;
    }

    private readonly bool IsEqual(BitVector Other)
    {
        return
            Data0 == Other.Data0 &&
            Data1 == Other.Data1 &&
            Data2 == Other.Data2 &&
            Data3 == Other.Data3;
    }


    public readonly bool Get(int Pos)
    {
        ToBitPos(Pos, out int PosInt, out int PosInInt);
        return GetBit(PosInt, PosInInt);
    }

    private readonly bool GetBit(int PosInt, int PosInInt)
    {
        fixed (int* IntPtr = &Data0)
        {
            return (IntPtr[PosInt] & (1 << PosInInt)) > 0;
        }
    }

    public void Set(int Pos, bool Value)
    {
        ToBitPos(Pos, out int PosByte, out int PosInByte);
        if (Value)
        {
            Set(PosByte, PosInByte);
        }
        else
        {
            Clear(PosByte, PosInByte);
        }
    }

    private void Set(int PosInt, int PosInInt)
    {
        fixed (int* IntPtr = &Data0)
        {
            IntPtr[PosInt] = (1 << PosInInt | IntPtr[PosInt]);
        }
    }

    private void Clear(int PosInt, int PosInInt)
    {
        fixed (int* IntPtr = &Data0)
        {
            IntPtr[PosInt] = (0 << PosInInt & IntPtr[PosInt]);
        }
    }

    private readonly void ToBitPos(int Pos, out int PosInt, out int PosInInt)
    {
        PosInt = Pos / IntSize;
        PosInInt = Pos - PosInt * IntSize;
    }

    public readonly int GetSelfIndexOf(Type Type)
    {
        int TargetIndex = ComponentAllocator.GetIDFor(Type);
        if (!Get(TargetIndex))
            return -1;

        int Count = 0;
        for (int i = 0; i < TargetIndex; i++)
        {
            Count += Get(i) ? 1 : 0;
        }
        return Count;
    }

    public readonly int[] GetSelfIndexOf(Type[] Types)
    {
        List<int> Result = new();
        foreach (var Type in Types)
        {
            int Value = GetSelfIndexOf(Type);
            if (Value == -1)
                continue;

            Result.Add(Value);
        }
        return Result.ToArray();
    }

    public readonly override string ToString()
    {
        StringBuilder SB = new(IntCount * IntSize);
        for (int i = 0; i < IntCount; i++)
        {
            for (int j = 0; j < IntSize; j++)
            {
                string Var = GetBit(i, j) ? "1" : "0";
                SB.Append(Var);
            }
        }
        return SB.ToString();
    }

    public readonly int GetFirstEmpty()
    {
        return GetFirst(0, false);
    }

    public readonly int GetFirst(int Start, bool bShouldBeSet)
    {
        int IntStart = Start / IntSize;
        int BitStart = Start - IntStart * IntSize;
        fixed (int* IntPtr = &Data0)
        {
            for (int i = IntStart; i < IntCount; i++)
            {
                int StartInInt = i == IntStart ? BitStart : 0;
                for (int j = StartInInt; j < IntSize; j++)
                {
                    var Temp = (IntPtr[i] & (1 << j)) >> j;
                    int Target = bShouldBeSet ? 1 : 0;
                    if (Temp != Target)
                        continue;

                    return i * IntSize + j;
                }
            }
        }
        return -1;
    }

    // see https://graphics.stanford.edu/~seander/bithacks.html
    public readonly int GetAmountSetBits()
    {
        int BitAmount = 0;
        fixed (int* IntPtr = &Data0)
        {
            for (int i = 0; i < IntCount; i++)
            {
                int Copy = IntPtr[i];
                Copy = Copy - ((Copy >> 1) & 0x55555555);
                Copy = (Copy & 0x33333333) + ((Copy >> 2) & 0x33333333);
                BitAmount += ((Copy + (Copy >> 4) & 0xF0F0F0F) * 0x1010101) >> 24;
            }
        }

        return BitAmount;
    }

    public readonly BitVector Clone()
    {
        BitVector Clone;
        Clone.Data0 = Data0;
        Clone.Data1 = Data1;
        Clone.Data2 = Data2;
        Clone.Data3 = Data3;
        return Clone;
    }

    public readonly override bool Equals(object obj)
    {
        if (obj is not BitVector Other)
            return false;

        return IsEqual(Other);
    }
    public readonly override int GetHashCode()
    {
        int Code = 0;
        Code += Data0.GetHashCode(); 
        Code += Data1.GetHashCode(); 
        Code += Data2.GetHashCode(); 
        Code += Data3.GetHashCode(); 
        return Code;
    }
}

File: Imports.Lib/src/EckyCS/Util/BitVector.cpp
#include "BitVector.h"

#include <assert.h>
#include <cstddef>
#include <memory>

#include "ComponentAllocator.h"

namespace EckyCS
{
    using namespace std;

    // see https://graphics.stanford.edu/~seander/bithacks.html
    int BitVector::GetAmountSetBits() const
    {
        auto Ptr = reinterpret_cast<const int*>(Data);
        int BitAmount = 0;
        for (int i = 0; i < IntCount; i++)
        {
            // make an actual copy to respect "const"
            auto Copy = *(Ptr + i);
            Copy = Copy - ((Copy >> 1) & 0x55555555);
            Copy = (Copy & 0x33333333) + ((Copy >> 2) & 0x33333333);
            BitAmount += ((Copy + (Copy >> 4) & 0xF0F0F0F) * 0x1010101) >> 24;
        }

        return BitAmount;
    }

    BitVector BitVector::And(const BitVector& Other) const
    {
        BitVector Result;
        for (int i = 0; i < ByteCount; i++)
        {
            Result.Data[i] = Data[i] & Other.Data[i];
        }
        return Result;
    }

    BitVector BitVector::Or(const BitVector& Other) const
    {
        BitVector Result;
        for (int i = 0; i < ByteCount; i++)
        {
            Result.Data[i] = Data[i] | Other.Data[i];
        }
        return Result;
    }

    BitVector BitVector::Subtract(const BitVector& Other) const
    {
        BitVector Result;
        for (int i = 0; i < ByteCount; i++)
        {
            int Temp = static_cast<int>(Data[i]) - static_cast<int>(Other.Data[i]);
            Result.Data[i] = static_cast<byte>(std::max(Temp, 0));
        }
        return Result;
    }

    void BitVector::Set(int Pos, bool Value)
    {
        assert(Pos >= 0 && Pos < TotalLength, "Out of bounds");
        int BytePos, PosInByte;
        ToBitPos(Pos, BytePos, PosInByte);
        if (Value)
        {
            SetBit(BytePos, PosInByte);
        }else
        {
            ClearBit(BytePos, PosInByte);
        }
    }

    bool BitVector::Get(int Pos) const
    {
        int BytePos, PosInByte;
        ToBitPos(Pos, BytePos, PosInByte);
        return GetBit(BytePos, PosInByte);
    }

    bool BitVector::operator==(const BitVector& Other) const
    {
        for (int i = 0; i < ByteCount; i++)
        {
            if (Data[i] != Other.Data[i])
                return false;
        }
        return true;
    }

    bool BitVector::operator!=(const BitVector& Other) const
    {
        return !(*this == Other);
    }

    int BitVector::GetSelfIndexOf(int Target) const
    {
        if (!Get(Target))
            return -1;

        int Count = 0;
        for (int i = 0; i < Target; ++i)
        {
            Count += Get(i);
        }
        return Count;
    }

    vector<int> BitVector::GetSelfIndicesOf(const vector<int>& Targets) const
    {
        vector<int> Result;
        Result.reserve(Targets.size());
        for (auto&& Target : Targets)
        {
            Result.emplace_back(GetSelfIndexOf(Target));
        }
        return Result;
    }

    int BitVector::GetFirstIndexWith(int Start, const bool bShouldBeSet) const
    {
        int ByteStart = Start / ByteSize;
        int BitStart = Start - ByteStart * ByteSize;

        // since we don't start at index 0, we have to skip some indices!
        for (int ByteIndex = ByteStart; ByteIndex < ByteCount; ++ByteIndex)
        {
            int StartInByte = ByteIndex == ByteStart ? BitStart : 0;
            for (int BitIndex = StartInByte; BitIndex < BitsPerByte; ++BitIndex)
            {
                // take only the important bit into consideration
                auto Temp = (0x1 & static_cast<int>(Data[ByteIndex] >> BitIndex)) > 0;
                if (Temp != bShouldBeSet)
                    continue;

                return ByteIndex * ByteSize + BitIndex;
            }
        }
        return -1;
    }

    vector<int> BitVector::GetAllIndicesWith(const bool bShouldBeSet) const
    {
        int Count = GetAmountSetBits();
        Count = bShouldBeSet ? Count : TotalLength - Count;
        vector<int> Result;
        Result.reserve(Count);

        for (int i = 0; i < TotalLength; ++i)
        {
            auto Temp = Get(i);
            if (Temp != bShouldBeSet)
                continue;

            Result.emplace_back(i);
        }
        return Result;
    }

    void BitVector::SetBit(int BytePos, int PosInByte)
    {
        Data[BytePos] |= static_cast<byte>(1 << PosInByte);
    }

    bool BitVector::GetBit(int BytePos, int PosInByte) const
    {
        return static_cast<int>(Data[BytePos] & static_cast<byte>(1 << PosInByte)) > 0;
    }

    void BitVector::ClearBit(int BytePos, int PosInByte)
    {
        Data[BytePos] &= static_cast<byte>(0 << PosInByte);
    }

    void BitVector::ToBitPos(int Pos, int& OutPosByte, int& OutPosBit)
    {
        OutPosByte = Pos / ByteSize;
        OutPosBit = Pos - OutPosByte * ByteSize;
    }
}
